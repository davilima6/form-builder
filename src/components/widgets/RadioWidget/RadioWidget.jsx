// @flow
import * as React from 'react';
import styled, { css } from 'styled-components';
import { Choice, Error, normalize } from '../../../utils';

const { useCallback } = React;

const WidgetWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;

  ${props => props.error
    && css`
      border-color: var(--dark-red);
    `};
`;

const Label = styled.label`
  flex-grow: 0.1;
`;

const StyledLabel = styled.div`
  flex: 0 0 100%;
  font-weight: 500;
  margin-bottom: 0.5em;
`;

const StyledChoice = styled.div`
  font-weight: 300;
  margin-bottom: 0.5em;
  text-transform: capitalize;
`;

const StyledInput = styled.input`
  font-size: 1em;
  padding: 3px 5px;
  width: 100%;
`;

const StyledError = styled.div`
  border: 1px dashed red;
`;

type Props = {
  id: string,
  label: string,
  choices: Array<Choice>,
  required?: boolean,
  value?: Choice,
  errors?: Array<Error>,
  onChangeField: Function,
};

const RadioWidget = ({
  id, label, choices, required, value, errors, onChangeField,
}: Props) => {
  const fieldId = `field-${id}`;

  const onChange = useCallback(
    ({ target }: SyntheticEvent<>) => {
      onChangeField(id, target.value);
    },
    [id, onChangeField],
  );

  return (
    <WidgetWrapper error={errors.length}>
      <StyledLabel>{label}</StyledLabel>
      {choices
        && choices.map((choice: Choice, enumKey: number) => {
          const fieldId = `field-${id}-${normalize(choice)}`;

          return (
            <Label htmlFor={fieldId} key={`${id}${choice}`}>
              <StyledChoice>{choice}</StyledChoice>
              <StyledInput
                type="radio"
                id={fieldId}
                name={id}
                value={enumKey}
                checked={value && enumKey === Number(value)}
                onChange={onChange}
              />
            </Label>
          );
        })}
      {errors.map(message => (
        <StyledError key={normalize(message)}>{message}</StyledError>
      ))}
    </WidgetWrapper>
  );
};
RadioWidget.defaultProps = {
  required: false,
  value: '',
  errors: [],
};

export default RadioWidget;
