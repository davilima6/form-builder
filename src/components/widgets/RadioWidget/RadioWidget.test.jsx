import React from 'react';
import { cleanup, render } from 'react-testing-library';
import 'jest-dom/extend-expect';
// import 'jest-styled-components';
import RadioWidget from './RadioWidget';

describe('RadioWidget', () => {
  let component;

  beforeEach(() => {
    const props = {
      label: 'email',
      type: 'radio',
      choices: ['Male', 'Female', 'Non-Binary'],
      data: { id: 0 },
      errors: [],
      onChange: () => {},
    };

    cleanup();
    component = render(<RadioWidget {...props} />);
  });

  test('renders without crashing', () => expect(component).toBeDefined());

  test.skip('matches snapshot', () => {
    const { asFragment } = component;

    expect(asFragment()).toMatchSnapshot();
    expect(asFragment()).toHaveStyleRule('background', '#fff');
  });
});
