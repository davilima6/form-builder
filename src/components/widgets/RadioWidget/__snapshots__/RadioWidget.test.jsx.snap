// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioWidget matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa djLaIN"
  >
    <label
      class="sc-bwzfXH gezcZz"
      for="field-0"
    >
      <span
        class="label"
      >
        yes
      </span>
      <input
        checked=""
        id="field-0"
        name="email"
        type="radio"
        value="0"
      />
    </label>
    <label
      class="sc-bwzfXH gezcZz"
      for="field-1"
    >
      <span
        class="label"
      >
        no
      </span>
      <input
        id="field-1"
        name="email"
        type="radio"
        value="1"
      />
    </label>
    <label
      class="sc-bwzfXH gezcZz"
      for="field-2"
    >
      <span
        class="label"
      >
        maybe
      </span>
      <input
        id="field-2"
        name="email"
        type="radio"
        value="2"
      />
    </label>
  </div>
</DocumentFragment>
`;
